File: rust/src/benchmarks/base64_encode.rs
use super::super::{Benchmark, helper};
use crate::config_i64;
use base64::{engine::general_purpose, Engine as _};

pub struct Base64Encode {
    n: i64,
    str_data: String,
    str2_encoded: String,
    result_val: u32,
}

impl Base64Encode {
    pub fn new() -> Self {
        let n = config_i64("Base64Encode", "size");

        let str_data = "a".repeat(n as usize);

        let str2_encoded = general_purpose::STANDARD.encode(&str_data);

        Self {
            n,
            str_data,
            str2_encoded,
            result_val: 0,
        }
    }
}

impl Benchmark for Base64Encode {
    fn name(&self) -> String {
        "Base64Encode".to_string()
    }

    fn run(&mut self, _iteration_id: i64) {

        let encoded = general_purpose::STANDARD.encode(&self.str_data);
        self.result_val = self.result_val.wrapping_add(encoded.len() as u32);
    }

    fn checksum(&self) -> u32 {
        let message = format!(
            "encode {} to {}: {}",
            if self.str_data.len() > 4 { format!("{}...", &self.str_data[0..4]) } else { self.str_data.clone() },
            if self.str2_encoded.len() > 4 { format!("{}...", &self.str2_encoded[0..4]) } else { self.str2_encoded.clone() },
            self.result_val
        );

        helper::checksum_str(&message)
    }
}

File: julia/src/benchmarks/Base64Encode.jl
using Base64

mutable struct Base64Encode <: AbstractBenchmark
    n::Int64
    str::String
    str2::String
    result::UInt32

    function Base64Encode()
        n = Helper.config_i64("Base64Encode", "size")
        new(n, "", "", UInt32(0))
    end
end

name(b::Base64Encode)::String = "Base64Encode"

function prepare(b::Base64Encode)
    b.str = "a" ^ b.n
    b.str2 = base64encode(b.str)
end

function run(b::Base64Encode, iteration_id::Int64)
    b.str2 = base64encode(b.str)
    b.result += UInt32(sizeof(b.str2))
end

function checksum(b::Base64Encode)::UInt32
    str_preview = length(b.str) > 4 ? b.str[1:4] * "..." : b.str
    str2_preview = length(b.str2) > 4 ? b.str2[1:4] * "..." : b.str2
    msg = "encode $(str_preview) to $(str2_preview): $(b.result)"
    return Helper.checksum(msg)
end
